docs/design.md

```diff
@@ -0,0 +1,339 @@
+}
+
+message UUID {
+     string value = 1;
+}
+
+
+```
+
+# <a name="_lmohbtf2asqg"></a>Authentication and Authorization
+Authentication will be handled using mutual TLS (mTLS).  Both the agent and the client will generate self signed certs using a common CA. When the connection is established, the client and agent will each verify the other’s authenticity.   Each client user will generate their own cert. The agent can retrieve the cert from gRPC connection and, since each cert is unique to a single user, the agent can distinguish which user executes which commands.  
+
+Authorization will ensure that a client user (as determined by the cert attached to the request) can only get the stop, get the status and stream the output of commands started by that user.  In other words, if user 1 starts a command, user 2 can’t stop, status or get the output of that command. Internally this will be accomplished by pairing the command execution with the public certificate attached to the request.
+
+# <a name="_tr2es7dqywgj"></a>mTLS
+Mutual TLS uses a dual exchange of certificates to verify the authenticity of both the host making the gRPC call and the host receiving the request.
+
+This  requires three files on each the agent and the client:
+
+1) The CA certificate file (shared between agent and client)
+1) The host’s self signed certificate file, which is unique to the agent and each client user 
+1) The host’s private key, which is also unique to the agent and each client user
+
+gRPC uses the CA certificate file to generate a certificate pool.  The self signed certificate and the private key are used to generate the certificate which will be presented in the authentication.
+
+By default, the[ TLS package](https://go.dev/src/crypto/tls/cipher_suites.go) only allows cipher suites with no known vulnerabilities. These are the encryption algorithms that this project will use.
+
+
+# <a name="_1y2ukim9xwfx"></a>Job Service
+
+The job service is a library responsible for managing jobs. The job service provides the interface to start, stop, status and get output of jobs.
+As each job is started, it sets up the cgroup by creating the directory and setting up the appropriate files. It then manages the lifecycle of the job, both notifying the job if the client stops the job and receiving notification as jobs complete. As the job changes state, the job service updates the job repository.
+
+The lifecycle of a job:
+
+1. The cgroup mounts, directories and files are setup
+1. The command is executed using exec.cmd Go library. The process is started in a goroutine then blocks waiting for Cmd.Wait() to return. When the command is started, the pid is added to cgroup.procs file and the Cmd object is saved in the JobService. The command is called through a wrapper script which ensures that the PID of the process is added to the procs file before execution begins (see cgroups section below for details)
+1. JobService updates that job's state in the repo to running

```

Comment:
Earlier we said that we have the following states:
```
     RUNNING = 0;
     COMPLETE = 1;
     STOPPED = 2;
     ERROR = 3;
```
But now, we are saying that we **update** to `running`.
So, we probably need an extra state (pending?) or do we intend to create the job with the `running` state initially?

Reply:
Added a pending state